File: top150/ArrayString/q121.java
package top150.ArrayString;

public class q121 {
    public int maxProfit(int[] prices) {
        int max = 0;
        int min = prices[0];

        for(int price: prices) {
            if(price - min > max) max = price-min;
            if(price < min) min = price;
        }

        return max;
    }

    public static void main(String[] args) {
        int[] prices = new int[]{7,1,5,3,6,4};

        int res = new q121().maxProfit(prices);
        System.out.println(res);
    }
}

File: top150/ArrayString/q121.py
import unittest


class Solution:
    def maxProfit(self, prices):
        n = len(prices)
        result = 0
        cur_min = prices[0]
        for p in prices[1:]:
            if p < cur_min:
                cur_min = p
            else:
                result = max(result, p-cur_min)

        return result


class Test(unittest.TestCase):
    def test_one(self):
        prices = [7, 1, 5, 3, 6, 4]
        sol = Solution()
        result = sol.maxProfit(prices)
        self.assertEqual(result, 5)

    def test_two(self):
        prices = [7, 6, 4, 3, 1]
        sol = Solution()
        result = sol.maxProfit(prices)
        self.assertEqual(result, 0)


# entry point
if __name__ == '__main__':
    unittest.main()
